src/test/java/org/springframework/integration/kafka/dsl/KafkaDslTests.java

```diff
@@ -259,16 +261,23 @@ public IntegrationFlow topic1ListenerFromKafkaFlow() {
 					.get();
 		}
 
+		@Bean
+		public ConcurrentKafkaListenerContainerFactory<Integer, String> kafkaListenerContainerFactory() {
+			ConcurrentKafkaListenerContainerFactory<Integer, String> factory =
+					new ConcurrentKafkaListenerContainerFactory<>();
+			factory.setConsumerFactory(consumerFactory());
+			factory.getContainerProperties().setAckMode(AckMode.MANUAL);
+			factory.setRecoveryCallback(new ErrorMessageSendingRecoverer(errorChannel(),
+					new RawRecordHeaderErrorMessageStrategy()));
+			factory.setRetryTemplate(new RetryTemplate());
+			return factory;
+		}
+
 		@Bean
 		public IntegrationFlow topic2ListenerFromKafkaFlow() {
 			return IntegrationFlows
-					.from(Kafka.messageDrivenChannelAdapter(consumerFactory(),
-							KafkaMessageDrivenChannelAdapter.ListenerMode.record, TEST_TOPIC2)
-							.configureListenerContainer(c ->
-									c.ackMode(ContainerProperties.AckMode.MANUAL))
-							.recoveryCallback(new ErrorMessageSendingRecoverer(errorChannel(),
-									new RawRecordHeaderErrorMessageStrategy()))
-							.retryTemplate(new RetryTemplate())
+					.from(Kafka.messageDrivenChannelAdapter(kafkaListenerContainerFactory().createContainer(TEST_TOPIC2),
```

Comment:
Sounds like you don't want to extend `Kafka` factory for fine-grained `KafkaListenerContainerFactory`-based methods :smile: 

Original request is like `Kafka.messageDrivenChannelAdapter(KafkaListenerContainerFactory, ListenerMode, String... topics)`

However I'm fine with your current decision on the matter - less code to support and less requests for extension, when we would have more `createContainer()` options.

Reply:
I started out with that approach but decided since we already have factory methods to inject a container, it would be simpler to do it this way, without losing much at all.

Reply:
+1 to stay with your choice here